### ISSUE TITLE
Memory Consumption for check --read-data

### ISSUE THREAD

**Issue by Radioactive-Man** [restic/restic#776]
restic 0.4.0 (v0.4.0-0-ga7c2f28)
compiled with go1.8rc3 on linux/arm

Compiled on windows with (set GOARM=7 and go run build.go --goos linux --goarch arm)

On my Banana Pi Pro with 1 GB of RAM the memory consumption while executing check --read-data on a repo with 65GB of data the memory usage increased from 700 MB to over 900 MB.

I traced it down to the following commit
```
212936eb529fc63c052721de20e379d9a81ec11f is the first bad commit
commit 212936eb529fc63c052721de20e379d9a81ec11f
Author: Alexander Neumann <alexander@bumpern.de>
Date:   Mon Jan 23 16:07:39 2017 +0100

    Make backend.LoadAll() similar to ioutil.ReadAll()

:040000 040000 539422371929cae18c62ab53ad3e6a2f3b4ad386 cc334a979bfe73dbac25a7b4aa28e4c0352ba04c M      src
```
Startig with this commit restic gets killed by the lowmemorykiller.

To force swapfile useage I need to set
`sysctl vm.swappiness=100`

**Comment by fd0** [comment]
Thanks for opening this issue, I'll look into it.

**Comment by fd0** [comment]
Could you please try the code in the branch `benchmark-checker` (see PR #779) and report back?

**Comment by Radioactive-Man** [comment]
Hi,

let it run for 45 min (whole check needs 4-5 hours) the output looks really good this is a dramatic reduce in memory consumption from 900MB down to 210MB this is even better than in version 0.3.0

I do a complete check of my repo this night so I can tell you tomorrow if something went wrong for the whole run.

```
root@bananapi ~ # htop

  1  [||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||100.0%]     Tasks: 32, 48 thr; 4 running
  2  [|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||94.5%]     Load average: 2.89 3.04 2.97
  Mem[|||||||||||||||||||||||||||||||||||||||||||||||||||||||||||276/970MB]     Uptime: 24 days, 05:03:15
  Swp[|||                                                         20/511MB]

  PID USER      PRI  NI  VIRT   RES   SHR S CPU% MEM%   TIME+  Command
10718 root       20   0  6560  2028  1436 S  0.0  0.2  0:00.68 -zsh
10942 root       20   0  6560  1436  1432 S  0.0  0.1  0:00.94 -zsh
10960 root       20   0  778M  209M  3248 S 187. 21.6  1h38:50 ./restic_arm -r ../Restic_Repo check --read-data
11001 root       20   0  778M  209M  3248 R 93.3 21.6  9:21.33 ./restic_arm -r ../Restic_Repo check --read-data
10970 root       20   0  778M  209M  3248 S 13.8 21.6  6:02.18 ./restic_arm -r ../Restic_Repo check --read-data
10989 root       20   0  778M  209M  3248 S 31.7 21.6  5:31.05 ./restic_arm -r ../Restic_Repo check --read-data
10965 root       20   0  778M  209M  3248 R 47.4 21.6  8:39.62 ./restic_arm -r ../Restic_Repo check --read-data
10961 root       20   0  778M  209M  3248 R  0.5 21.6  0:32.60 ./restic_arm -r ../Restic_Repo check --read-data
10971 root       20   0  778M  209M  3248 S  0.0 21.6  6:48.88 ./restic_arm -r ../Restic_Repo check --read-data
10985 root       20   0  778M  209M  3248 S  0.0 21.6  9:28.37 ./restic_arm -r ../Restic_Repo check --read-data
10962 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.22 ./restic_arm -r ../Restic_Repo check --read-data
10963 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10964 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10966 root       20   0  778M  209M  3248 S  0.0 21.6  0:04.91 ./restic_arm -r ../Restic_Repo check --read-data
10967 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.77 ./restic_arm -r ../Restic_Repo check --read-data
10969 root       20   0  778M  209M  3248 S  0.0 21.6  5:53.80 ./restic_arm -r ../Restic_Repo check --read-data
10972 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.10 ./restic_arm -r ../Restic_Repo check --read-data
10973 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.08 ./restic_arm -r ../Restic_Repo check --read-data
10974 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.07 ./restic_arm -r ../Restic_Repo check --read-data
10975 root       20   0  778M  209M  3248 S  0.0 21.6  0:10.06 ./restic_arm -r ../Restic_Repo check --read-data
10976 root       20   0  778M  209M  3248 S  0.0 21.6 11:04.25 ./restic_arm -r ../Restic_Repo check --read-data
10977 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10978 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.01 ./restic_arm -r ../Restic_Repo check --read-data
10979 root       20   0  778M  209M  3248 S  0.0 21.6  0:04.40 ./restic_arm -r ../Restic_Repo check --read-data
10980 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.01 ./restic_arm -r ../Restic_Repo check --read-data
10981 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.01 ./restic_arm -r ../Restic_Repo check --read-data
10982 root       20   0  778M  209M  3248 S  0.0 21.6  0:01.14 ./restic_arm -r ../Restic_Repo check --read-data
10983 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10984 root       20   0  778M  209M  3248 S  0.0 21.6  0:01.04 ./restic_arm -r ../Restic_Repo check --read-data
10986 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10987 root       20   0  778M  209M  3248 S  0.0 21.6  0:01.32 ./restic_arm -r ../Restic_Repo check --read-data
10988 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10990 root       20   0  778M  209M  3248 S  0.0 21.6  0:01.44 ./restic_arm -r ../Restic_Repo check --read-data
10991 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.04 ./restic_arm -r ../Restic_Repo check --read-data
10992 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.01 ./restic_arm -r ../Restic_Repo check --read-data
10993 root       20   0  778M  209M  3248 S  0.0 21.6  0:01.30 ./restic_arm -r ../Restic_Repo check --read-data
10994 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10995 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.08 ./restic_arm -r ../Restic_Repo check --read-data
10996 root       20   0  778M  209M  3248 S  0.0 21.6 10:17.61 ./restic_arm -r ../Restic_Repo check --read-data
10997 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
10998 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.05 ./restic_arm -r ../Restic_Repo check --read-data
10999 root       20   0  778M  209M  3248 S  0.0 21.6  6:36.25 ./restic_arm -r ../Restic_Repo check --read-data
11000 root       20   0  778M  209M  3248 S  0.0 21.6  6:42.18 ./restic_arm -r ../Restic_Repo check --read-data
11002 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
11003 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
11004 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.04 ./restic_arm -r ../Restic_Repo check --read-data
11005 root       20   0  778M  209M  3248 S  0.0 21.6  0:01.04 ./restic_arm -r ../Restic_Repo check --read-data
11006 root       20   0  778M  209M  3248 S  0.0 21.6  0:00.00 ./restic_arm -r ../Restic_Repo check --read-data
 1539 root       20   0  1792     4     0 S  0.0  0.0  0:00.01 /sbin/getty -L ttyS0 115200 vt100
 1533 root       20   0  3544     4     0 S  0.0  0.0  0:00.01 /sbin/getty 38400 tty1
 1534 root       20   0  3544     4     0 S  0.0  0.0  0:00.01 /sbin/getty 38400 tty2
F1Help  F2Setup F3SearchF4FilterF5Tree  F6SortByF7Nice -F8Nice +F9Kill  F10Quit
```

**Comment by Radioactive-Man** [comment]
Complete run finished without problems everything fine

**Comment by fd0** [comment]
Cool, then I'll close this issue as resolved by #779 (I've merged that PR a moment ago).
